torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,401 @@
+        _cache_enabled -= 1
+        if not _cache_enabled:
+            _cache = dict.fromkeys(_cache, None)
+
+
+class Parametrization(Module):
+    r"""A kind of Module that parametrizes a Parameter or a buffer in terms
+    of its forward function
+
+    After registering a Parametrization on a tensor ``t`` within a module with :meth:`
+    #register_parametrization`, ``module.t`` will be turn into a property, and will
+    return ``parametrization(t)``, where ``t`` is the previous tensor
+
+    After registering a Parametrization, the parametrization will be accessible under
+    ``module.parametrizations.t``. The unparametrized tensor may be accessed through
+    ``parametrization.original_tensor()``
+
+    Before a parametrization is registered, it may be chained to others to compose them
+    via the ``chain`` method.
+
+    Parametrizations may be registered on parameters, buffers, and already parametrized
+    tensors
+
+        # Defining several parametrizations on the same tensor
+        class MyParametrization(nn.Parametrization):
+            def forward(self, X):
+                return 2. * X
+        module = nn.Linear(5,3)
+        p1 = MyParametrization()
+        p2 = MyParametrization()
+        # Register p1 on the Parameter "weight"
+        register_parametrization(module, "weight", p1)
+        # Register p2 on the parametrized Parameter "weight"
+        register_parametrization(module, "weight", p2)
+        assert(module.weight == 4. * module.parametrizations.weight.original_tensor())))
+
+    We can also first compose them and then register them. The following code is
+    equivalent to the previous one
+
+        module = nn.Linear(5, 3)
+        # Chain the two parametrizations and register them
+        p2.chain(p1)
+        register_parametrization(module, "weight", p2)
+        # Now p2.evaluate() computes p1 and then p2
+        assert(module.weight == 4. * module.parametrizations.weight.original_tensor())))
+
+
+    Parametrized parameters and buffers have an in-built caching system via the context
+    manager :class:``cached``
+    """
+
+    def evaluate(self):
+        r"""Evaluates the parametrization"""
+        if self.is_registered():
+            return self(self.original)
+        else:
+            raise RuntimeError(
+                "A Parametrization cannot be evaluated before " "registering it"
+            )
+
+    def original_tensor(self):
+        r"""Returns the tensor the parametrization was registered on"""
+        try:
+            return self.last_parametrization().original
+        except AttributeError:
+            raise ValueError("The parametrization is not registered to a tensor.")
+
+    def last_parametrization(self):
+        r"""Returns the last parametrization of the chain.
+
+        In particular, if the parametrization is not chained, it returns itself
+        """
+        last = self
+        while last.is_chained():
+            last = last.parametrizations.original
```

Comment:
I'm a bit confused. I thought that we've decided that in the end we shouldn't make parametrizations aware of the chain they are in. Are you still planning to remove this?